components/builder-graph/src/graph_helpers.rs

```diff
@@ -428,6 +437,9 @@ pub fn transitive_deps(graph: &DiGraphMap<PackageIdentIntern, EdgeType>,
                 EdgeType::RuntimeDep => true,
                 EdgeType::BuildDep => include_build_deps,
                 EdgeType::StrongBuildDep => include_build_deps,
+                EdgeType::ExternalConstraint => {
+                    unimplemented!("External Constraints should not appear here")
```

Comment:
Similar comment here as the comment for the previous `unimplemented!()`. Whether this remains, or becomes a warning message in the logs, it should probably have a unique message so we can distinguish and easily identify exactly where the `EdgeType::ExternalConstraint` violation happened.

Reply:
Unlike the graph dumping code above, if we hit this here we have a deeply, terribly broken graph. Ideally we'd represent this as a different type to try and forbid it totally, but due to the way we reuse some code across phases it would require some serious refactoring.

Reply:
I see, it sounds like we actually want to crash here then because we're in an unknown state.

In that case, I think replacing `unimplemented!()` with `panic!()` is going to serve the reader better as we have no intention of eventually implementing something for this match branch arm.

Reply:
Agreed